src/semaphore.c

```diff
@@ -804,3 +889,110 @@ _dispatch_thread_semaphore_wait(_dispatch_thread_semaphore_t sema)
 #error "No supported semaphore type"
 #endif
 }
+
+#if USE_FUTEX_SEM
+#define DISPATCH_FUTEX_NUM_SPINS 100
+#define DISPATCH_FUTEX_VALUE_MAX INT_MAX
+#define DISPATCH_FUTEX_NWAITERS_SHIFT 32
+#define DISPATCH_FUTEX_VALUE_MASK ((1ull << DISPATCH_FUTEX_NWAITERS_SHIFT) - 1)
+
+static bool _dispatch_futex_wait_slow(dispatch_futex_t dfx,
+		const struct timespec* timeout);
+
+DISPATCH_INLINE
+static int
+_dispatch_futex_syscall(int *futex_addr, 
+		int op, int val, const struct timespec *timeout)
+{
+	return (int)syscall(SYS_futex, futex_addr, op | FUTEX_PRIVATE_FLAG, val, 
+						timeout, NULL, 0);
+}
+
+/* all memory orders below were added with strict seq_cst semantics
+ * assumed to be the strongest
+ */
+
+// Atomic-decrement-if-positive on the semaphore value, with acquire MO if
+// successful.
+DISPATCH_INLINE
+bool
+_dispatch_futex_trywait(dispatch_futex_t dfx)
+{
+	uint64_t orig;
+	while ((orig = dfx->dfx_data) & DISPATCH_FUTEX_VALUE_MASK) {
+		if (dispatch_atomic_cmpxchg2o(dfx, dfx_data, orig, orig - 1, acquire)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+// Increments semaphore value and reads wait count in a single atomic
+// operation, with release MO. If wait count is nonzero, issues a FUTEX_WAKE.
+DISPATCH_INLINE
+bool
+_dispatch_futex_signal(dispatch_futex_t dfx)
+{
+	uint64_t orig;
+	do {
+		orig = dfx->dfx_data;
+		if (slowpath((orig & DISPATCH_FUTEX_VALUE_MASK) ==
+					 DISPATCH_FUTEX_VALUE_MAX)) {
+			DISPATCH_CRASH("semaphore overflow");
+		}
+	} while (!dispatch_atomic_cmpxchgvw2o(dfx, dfx_data, orig, orig + 1, &orig, release));
+	if (slowpath(orig >> DISPATCH_FUTEX_NWAITERS_SHIFT)) {
+		int ret = _dispatch_futex_syscall(&dfx->dfx_futex, FUTEX_WAKE, 1, NULL);
```

Comment:
ok I'm wrong the man page says:

```
       EINTR  A FUTEX_WAIT operation was interrupted by  a  signal  (see  sig‐
              nal(7)) or a spurious wakeup.
```

so that is fine